File: src/readssidev.c
/*
 * NAME:
 * readssidev.c
 *
 * PURPOSE:
 * To dump data stored in NCSA HDF5 files by software collocate...
 *
 * REQUIREMENTS:
 * NA
 *
 * INPUT:
 * NA
 *
 * OUTPUT:
 * NA
 *
 * NOTES:
 * The same files as for DLI are used, thus the DLIDEV keyword is used
 * herein.
 *
 * BUGS:
 * NA
 *
 * AUTHOR:
 * Øystein Godøy, MET/FOU, 21.01.2003
 *
 * MODIFIED:
 * NA
 *
 * CVS_ID:
 * $Id: readssidev.c,v 1.1 2007-09-07 17:03:00 steingod Exp $
 */

#ifdef HAVE_LIBHDF5

#include <readdlidev.h>
#include <unistd.h>
#include <dirent.h>
#include <time.h>
#include <utime.h>
#include <sys/types.h>
#include <sys/stat.h>

/*
int sm_debug=1;
*/

void readssidev(char **infile, int *n, int *p, char **station, 
	int *start, int *end,
	int *mytime,
	double *k1, double *k2, 
	double *k3a, double *k3b, 
	double *k4, double *k5,
	double *soz, double *saz, double *raz,
	double *t2m, double *ps, double *pw,
	int *cm,
	double *ssi) {

    /* General variables */
    char *where="readssidev";
    int i, j, k;
    int status, noobs;
    as_data *asdata;
    ns_data *nsdata;
    safcm_data2 *cmdata;
    safssi_data *ssidata;
    skeys scrit = {"DLIDEV","NA",0,0};
    struct tm ut;
    time_t vt, mymax=0, mymin=INT_MAX;

    if (*p != BOXSIZE2D) {
	error(where,"BOXSIZE differs...");
	return;
    }

    sprintf(scrit.station,"%s",*station);
    scrit.t_start = (time_t) *start;
    scrit.t_end = (time_t) *end;
    printf(">> %d %d\n",(int) scrit.t_start,(int) scrit.t_end);

    if ((noobs=read1245(*infile, &scrit, 
		    &asdata, &nsdata, &cmdata, &ssidata)) <= 0) {
	error(where," File trouble...");
    }
    if (noobs != *n) {
	error(where,"Allocated space differs...");
	printf(" %d %d\n", noobs, *n);
	return;
    }

    /*
     * Retrieve data for analysis within R
     */
    printf(" noobs: %d, p: %d\n", noobs, *p);
    j=0;
    for (k=0;k<noobs;k++) {
	for (i=0;i<*p;i++) {
	    mytime[j] = (int) asdata[k].vtime;
	    k1[j] = (double) asdata[k].ch1[i];
	    k2[j] = (double) asdata[k].ch2[i];
	    k3a[j] = (double) asdata[k].ch3a[i];
	    k3b[j] = (double) asdata[k].ch3b[i];
	    k4[j] = (double) asdata[k].ch4[i];
	    k5[j] = (double) asdata[k].ch5[i];
	    soz[j] = (double) asdata[k].ang.soz;
	    saz[j] = (double) asdata[k].ang.saz;
	    raz[j] = (double) asdata[k].ang.raz;
	    t2m[j] = (double) nsdata[k].t2m[0];
	    ps[j] = (double) nsdata[k].ps[0];
	    pw[j] = (double) nsdata[k].pw[0];
	    cm[j] = (int) cmdata[k].data[i];
	    ssi[j] = (double) ssidata[k].data[i];
	    j++;
	}
    }
    /*
    printf(" Max utime: %d\n Min utime: %d\n",(int) mymax,(int) mymin);
    */

    return;
}

#endif

File: R/readssidev.R
#
# NAME:
# readssidev
#
# PURPOSE:
# To read collocate HDF5 files containing SSI development database.
#
# NOTES:
# Not fully finished... 
#
# BUGS:
# Probably in the underlying C-function...
# C functions does not transfer strings back properly...
#
# AUTHOR:
# Øystein Godøy, MET/FOU, 30.01.2003
#
# CVS_ID:
# $Id: readssidev.R,v 1.2 2013-04-11 20:29:04 steingod Exp $
#

readssidev <- function(filename,
    station="NA",start="NA",end="NA",classname="DLIDEV") {

    if (missing(filename)) {
	cat("Husk at filnavn må oppgis...\n")
	return;
    }

    start <- as.POSIXct(strptime(start,format="%d%b%Y"))
    end <- as.POSIXct(strptime(end,format="%d%b%Y"))
    if (is.na(start)) start <- 0
    if (is.na(end)) end <- 0

    noobs <- 0

    cat("Sjekker antall poster som skal leses...\n")
    tmp <- .C("checkrecdli",
	filename=as.character(filename),
	noobs=as.integer(noobs),
	classname=as.character(classname),
	station=as.character(station),
	start=as.integer(start),end=as.integer(end))
    if (tmp$noobs <= 0) return

    cat(paste("Antall poster funnet:",tmp$noobs,"\n"))

    nopix <- 13*13
    size <- tmp$noobs*nopix
    time <- vector(mode="integer",length=size)
    k1 <- real(length=size)
    k2 <- real(length=size)
    k3a <- real(length=size)
    k3b <- real(length=size)
    k4 <- real(length=size)
    k5 <- real(length=size)
    soz <- real(length=size)
    saz <- real(length=size)
    raz <- real(length=size)
    t2m <- real(length=size)
    pw <- real(length=size)
    ps <- real(length=size)
    cm <- vector(mode="integer",length=size)
    ssi <- real(length=size)
    
    cat("Leser poster...\n")
    cat(paste("Har satt av:",size,"elementer\n"))
    tmp <- .C("readdlidev",
	filename=as.character(filename),
	noobs=as.integer(tmp$noobs),nopix=as.integer(nopix),
	station=as.character(station),
	start=as.integer(start),end=as.integer(end),
	time=as.integer(time),
	k1=as.real(k1),k2=as.real(k2),
	k3a=as.real(k3a),k3b=as.real(k3b),
	k4=as.real(k4),k5=as.real(k5),
	soz=as.real(soz),saz=as.real(saz),raz=as.real(raz),
	t2m=as.real(t2m),ps=as.real(ps),pw=as.real(pw),
	cm=as.integer(cm),
	ssi=as.real(ssi))

    return(tmp)
}
